File: sdk/elixir/lib/experimentation_platform/cache.ex
defmodule ExperimentationPlatform.Cache do
  @moduledoc """
  ETS-backed GenServer cache for server results (flag evaluations and
  experiment assignments), keyed per user + key.

  Provides a process-safe, TTL-aware in-memory cache with a configurable
  maximum size. When the cache reaches `max_cache_size`, the oldest inserted
  entry is evicted to make room for a new one.

  ## Storage Format

  Each ETS entry is stored as `{key, value, expires_at_ms, seq}` where
  `expires_at_ms` is a monotonic millisecond timestamp and `seq` a strictly
  increasing insertion counter (used for ordering and eviction).

  The client uses tuple keys so entries can be listed per user:

  - `{:flag, user_id, flag_key}` -> `%ExperimentationPlatform.FlagEvaluation{}`
  - `{:assignment, user_id, experiment_key}` -> `%ExperimentationPlatform.Assignment{}`

  Writes are synchronous calls, so a `put/4` is complete before it returns.

  ## Usage

      {:ok, cache} = Cache.start_link_unnamed(config)
      :ok = Cache.put(cache, {:flag, "user-1", "my-flag"}, evaluation)
      Cache.get(cache, {:flag, "user-1", "my-flag"})
      Cache.list(cache, :flag, "user-1")
  """

  use GenServer

  alias ExperimentationPlatform.Config

  @table_name :experimentation_platform_cache

  # --- Client API ---

  @doc "Start the Cache GenServer registered under the module name."
  @spec start_link(Config.t()) :: GenServer.on_start()
  def start_link(config) do
    GenServer.start_link(__MODULE__, config, name: __MODULE__)
  end

  @doc "Start a Cache GenServer without registering a global name (one per client)."
  @spec start_link_unnamed(Config.t()) :: GenServer.on_start()
  def start_link_unnamed(config) do
    GenServer.start_link(__MODULE__, config)
  end

  @doc "Retrieve a cached value. Returns `nil` if missing or expired."
  @spec get(GenServer.server(), any()) :: any()
  def get(server \\ __MODULE__, key) do
    GenServer.call(server, {:get, key})
  end

  @doc """
  Store a value with an optional TTL override in seconds
  (`config.cache_ttl` when `nil`). Synchronous.
  """
  @spec put(GenServer.server(), any(), any(), non_neg_integer() | nil) :: :ok
  def put(server \\ __MODULE__, key, value, ttl_seconds \\ nil) do
    GenServer.call(server, {:put, key, value, ttl_seconds})
  end

  @doc """
  Live (non-expired) values whose key is `{type, user_id, _}`, oldest first.
  Expired entries encountered on the way are pruned.
  """
  @spec list(GenServer.server(), atom(), String.t()) :: [any()]
  def list(server \\ __MODULE__, type, user_id) do
    GenServer.call(server, {:list, type, user_id})
  end

  @doc "Delete a specific cache entry. Synchronous."
  @spec delete(GenServer.server(), any()) :: :ok
  def delete(server \\ __MODULE__, key) do
    GenServer.call(server, {:delete, key})
  end

  @doc "Clear all cache entries. Synchronous."
  @spec clear(GenServer.server()) :: :ok
  def clear(server \\ __MODULE__) do
    GenServer.call(server, :clear)
  end

  @doc "Number of entries currently stored (including expired ones not yet pruned)."
  @spec size(GenServer.server()) :: non_neg_integer()
  def size(server \\ __MODULE__) do
    GenServer.call(server, :size)
  end

  # --- GenServer Callbacks ---

  @impl true
  def init(%Config{} = config) do
    table = :ets.new(@table_name, [:set, :private])
    {:ok, %{table: table, config: config, seq: 0}}
  end

  @impl true
  def handle_call({:get, key}, _from, %{table: table} = state) do
    {:reply, lookup(table, key), state}
  end

  def handle_call({:put, key, value, ttl_override}, _from, state) do
    %{table: table, config: config, seq: seq} = state
    ttl = ttl_override || config.cache_ttl
    expires_at = mono_ms() + ttl * 1_000

    unless :ets.member(table, key) do
      evict_if_needed(table, config.max_cache_size)
    end

    :ets.insert(table, {key, value, expires_at, seq})
    {:reply, :ok, %{state | seq: seq + 1}}
  end

  def handle_call({:list, type, user_id}, _from, %{table: table} = state) do
    now = mono_ms()

    values =
      table
      |> :ets.match_object({{type, user_id, :_}, :_, :_, :_})
      |> Enum.filter(fn {key, _value, expires_at, _seq} ->
        if now < expires_at do
          true
        else
          :ets.delete(table, key)
          false
        end
      end)
      |> Enum.sort_by(fn {_key, _value, _expires_at, seq} -> seq end)
      |> Enum.map(fn {_key, value, _expires_at, _seq} -> value end)

    {:reply, values, state}
  end

  def handle_call({:delete, key}, _from, %{table: table} = state) do
    :ets.delete(table, key)
    {:reply, :ok, state}
  end

  def handle_call(:clear, _from, %{table: table} = state) do
    :ets.delete_all_objects(table)
    {:reply, :ok, state}
  end

  def handle_call(:size, _from, %{table: table} = state) do
    {:reply, :ets.info(table, :size), state}
  end

  # --- Private Helpers ---

  defp lookup(table, key) do
    case :ets.lookup(table, key) do
      [{^key, value, expires_at, _seq}] ->
        if mono_ms() < expires_at do
          value
        else
          :ets.delete(table, key)
          nil
        end

      [] ->
        nil
    end
  end

  defp evict_if_needed(table, max_size) do
    if :ets.info(table, :size) >= max_size do
      # Evict the oldest inserted entry (smallest seq)
      oldest =
        :ets.foldl(
          fn {key, _value, _expires_at, seq}, acc ->
            case acc do
              nil -> {seq, key}
              {min_seq, _} when seq < min_seq -> {seq, key}
              _ -> acc
            end
          end,
          nil,
          table
        )

      case oldest do
        {_seq, key} -> :ets.delete(table, key)
        nil -> :ok
      end
    end
  end

  defp mono_ms do
    System.monotonic_time(:millisecond)
  end
end

File: sdk/ruby/lib/experimentation_platform/cache.rb
module ExperimentationPlatform
  # Thread-safe TTL cache with LRU eviction.
  #
  # Entries are stored as { value:, expires_at: } pairs.
  # The insertion order of @store (Hash in Ruby >= 1.9) is used for LRU eviction:
  # the first entry is the oldest.
  #
  # Keys may be any hashable object; the client uses Array keys such as
  # [:flag, user_id, flag_key] so entries can be listed per user.
  class Cache
    # @param ttl          [Integer] default time-to-live in seconds
    # @param max_size     [Integer] maximum number of entries; oldest evicted when exceeded
    def initialize(ttl: 300, max_size: 1000)
      @default_ttl = ttl.to_i
      @max_size    = max_size.to_i
      @store       = {}          # key => { value:, expires_at: }
      @mutex       = Mutex.new
    end

    # Retrieve a cached value.
    #
    # @param key [Object]
    # @return [Object, nil] the cached value, or nil if missing or expired
    def get(key)
      @mutex.synchronize do
        entry = @store[key]
        return nil if entry.nil?

        if Time.now >= entry[:expires_at]
          @store.delete(key)
          return nil
        end

        # Refresh LRU position
        @store.delete(key)
        @store[key] = entry

        entry[:value]
      end
    end

    # Store a value in the cache.
    #
    # @param key   [Object]
    # @param value [Object]
    # @param ttl   [Integer, nil] seconds; uses default_ttl when nil
    def set(key, value, ttl: nil)
      effective_ttl = (ttl || @default_ttl).to_i
      expires_at    = Time.now + effective_ttl

      @mutex.synchronize do
        # Remove existing entry to refresh LRU position
        @store.delete(key)

        # Evict oldest entry if at capacity
        if @store.size >= @max_size
          oldest_key = @store.keys.first
          @store.delete(oldest_key)
        end

        @store[key] = { value: value, expires_at: expires_at }
      end

      value
    end

    # Values of all live (non-expired) entries whose key satisfies the block,
    # in insertion order. Expired entries encountered on the way are pruned.
    # The block receives the key only and must not call back into the cache.
    #
    # @yieldparam key [Object]
    # @return [Array<Object>]
    def live_values
      now = Time.now
      @mutex.synchronize do
        expired = []
        values  = []
        @store.each do |key, entry|
          next unless yield(key)

          if now >= entry[:expires_at]
            expired << key
          else
            values << entry[:value]
          end
        end
        expired.each { |key| @store.delete(key) }
        values
      end
    end

    # Remove a specific entry.
    #
    # @param key [Object]
    def delete(key)
      @mutex.synchronize { @store.delete(key) }
    end

    # Remove all entries.
    def clear
      @mutex.synchronize { @store.clear }
    end

    # Return the number of entries currently stored (including expired ones not yet evicted).
    #
    # @return [Integer]
    def size
      @mutex.synchronize { @store.size }
    end
  end
end
